src/api/app/models/worker_status.rb

```diff
@@ -1,25 +1,45 @@
 class WorkerStatus
-  def self.hidden
-    mydata = Rails.cache.read('workerstatus')
-    ws = Nokogiri::XML(mydata || Backend::Api::BuildResults::Worker.status).root
-    prjs = {}
-    ws.css('building').each do |b|
-      prjs[b['project']] = 1
+  class << self
+    def hidden
+      mydata = Rails.cache.fetch('workerstatus') do
+        Backend::Api::BuildResults::Worker.status
+      end
+      # remove information about projects which are not visible to the current user
+      hide_project_information(Nokogiri::XML(mydata).root)
+    end
+
+    private
+
+    def hide_project_information(worker_status_root)
+      worker_status = worker_status_root
+      prjs = initialize_projects(worker_status)
+      xpath_string = generate_xpath(project_names(prjs.keys))
+      worker_status.xpath(xpath_string).each do |b|
+        b['project'] = '---'
+        b['repository'] = '---'
+        b['package'] = '---'
+      end
+      worker_status
     end
-    names = {}
-    # now try to find those we have a match for (the rest are hidden from you
-    Project.where(name: prjs.keys).pluck(:name).each do |n|
-      names[n] = 1
+
+    # use Xpath to filter the projects
+    def generate_xpath(names)
+      return '//building' if names.empty?
+      "//building[not(@project=#{xpath_list(names)})]"
     end
-    ws.css('building').each do |b|
-      # no prj -> we are not allowed
-      next if names.key?(b['project'])
-      Rails.logger.debug "workerstatus2clean: hiding #{b['project']} for user #{User.possibly_nobody.login}"
-      b['project'] = '---'
-      b['repository'] = '---'
-      b['package'] = '---'
+
+    # from array generates a xpath list like ('name1', 'name2', 'nameX')
+    def xpath_list(names)
+      format("(#{names.collect { |_f| '\'%s\'' }.join(',')})", *names)
```

Comment:
I would have expected some higher level of paranoia from you when it comes to quoting :)

Reply:
:)

I assumed that because there is no user input involved and it will actually handle well-formatted project name, it would be ok.. argh this "_f"  :angry: 

I would love to have some help to improve this code, I failed to find some ruby magic one-liner to transform an array of strings to this XPATH sequence. Any idea @coolo ?

Reply:
Paranoia about what? That the xpath query fails? :scream: